gedi/process_l2a.py

```diff
@@ -0,0 +1,290 @@
+import argparse
+import h5py
+import os
+import numpy as np
+import pandas as pd
+import geopandas as gp
+
+from typing import List
+from shapely.geometry import Point
+
+
+def gedi_L2A_to_df(
+    file_dir: str,
+    file_paths: List[str],
+    bbox: List[float],
+    layers: List[str] = [
+        "shot_number",
+        "lat_lowestmode",
+        "lon_lowestmode",
+        "elev_lowestmode",
+        "elev_highestreturn",
+        "sensitivity",
+        "quality_flag",
+        "rh",
+    ],
+    latlayer: str = "lat_lowestmode",
+    lonlayer: str = "lon_lowestmode",
+) -> pd.DataFrame:
+    """
+    Given the absolute path to a directory of GEDI Level 2A h5 files, the
+    names of the h5 files, and information about the desired output, ingest
+    the h5 files and output a pd.DataFrame containing data for all the valid
+    shots within the bounding box.
+
+    The output will have an array-valued column that is designed to be handled
+    by append_canopy_metrics and then deleted from the DataFrame.
+
+    Parameters
+    ----------
+    file_dir : str
+        Path to directory containing the h5 files, including a trailing slash
+        such that {file_dir}{fname} is a valid path to a file, given fname a
+        valid file name.
+    file_paths : List[str]
+        The names of the h5 files to process.
+    bbox : List[float]
+        The coordinates of the bounding box, formatted as
+        [ul_lat, ul_lon, lr_lat, lr_lon].
+    layers : List[str]
+        The columns for the output DataFrame.
+    latlayer : str
+        The name of the latitude layer.
+    lonlayer : str
+        The name of the longitude layer.
+
+    Returns
+    -------
+    pd.DataFrame containing the user-specified contents of the h5 files for the
+    valid shots within the specified bounding box.
+    """
+    df = pd.DataFrame()
+    for _f_name in file_paths:
+        try:
+            _f = h5py.File(f"{file_dir}{_f_name}", "r")
+            print(f"Processing file {_f_name}")
+            [ul_lat, ul_lon, lr_lat, lr_lon] = bbox
+            for beam in [
+                "BEAM0000",
+                "BEAM0001",
+                "BEAM0010",
+                "BEAM0011",
+                "BEAM0101",
+                "BEAM0110",
+                "BEAM1000",
+                "BEAM1011",
+            ]:
+                tmp_df = pd.DataFrame()
+                x = _f[beam][latlayer][()]
+                y = _f[beam][lonlayer][()]
+                qual = _f[beam]["quality_flag"][()]
+
+                mask = (
+                    (qual == 1)
+                    & (x <= ul_lat)
+                    & (x >= lr_lat)
+                    & (y <= lr_lon)
+                    & (y >= ul_lon)
+                )
+                tmp_df["beam"] = [beam] * np.count_nonzero(mask)
+
+                for layer in layers:
+                    tmp_df[layer] = _f[beam][layer][()][mask].tolist()
+
+                df = df.append(tmp_df)
+
+        except KeyError as e:
```

Comment:
Initially, I decided to catch the KeyError here in order to get the rest of the processing to run. It seems to be happening because, perhaps due to coverage/data availability issues, a few files (7 of the 65 in our area of interest) were quite small and missing data, which would lead to a KeyError being thrown here. As far as I can tell, this is a NASA-side issue (i.e. not due to an incomplete download to my local machine). 

Especially open to a more robust or better suggestion for how to handle this.